File: shortest path in undirected graph.py
from collections import deque
class Solution:
    def shortestPath(self, edges, n, m, src):
        # code here
        adjList = [[] for _ in range(n)]
        for edge in edges:
            u, v = edge
            adjList[u].append(v)
            adjList[v].append(u)
        visited = [0 for _ in range(n)]
        q = deque()
        distance = [float('inf') for _ in range(n)]
        
        q.append((src, 0))
        visited[src] = 1
        distance[src] = 0
        
        while q:
            currentNode, currentNodeDistance = q.popleft()
            for adjacentNode in adjList[currentNode]:
                if not visited[adjacentNode]:
                    vDistance = currentNodeDistance + 1
                    if vDistance < distance[adjacentNode]:
                        distance[adjacentNode] = vDistance
                        visited[adjacentNode] = 1
                        q.append((adjacentNode, vDistance))
        for index, value in enumerate(distance):
            if value == float('inf'):
                distance[index] = -1
        return distance        
            

File: shortest path in undirected graph.cpp
class Solution {
  public:
    vector<int> shortestPath(vector<vector<int>>& edges, int N,int M, int src){
        // code here
        vector<vector<int>> adjList(N);
        for (auto &it: edges){
            adjList[it[0]].push_back(it[1]);
            adjList[it[1]].push_back(it[0]);
        }
        // every index indicates distance from source node
        vector<int> distance(N, 1e9);
        queue<int> q;
        distance[src] = 0;
        q.push(src);
        while (!q.empty()){
            int current = q.front();
            q.pop();
            for (auto &adjacentNode: adjList[current]){
                // distance from current node to adjacentNode
                int currentDistance = distance[current] + 1;
                if (currentDistance < distance[adjacentNode]){
                    distance[adjacentNode]= currentDistance;
                    q.push(adjacentNode);
                }
            }
        }
        for (int i =0; i < N; i++){
            if (distance[i] == 1e9){
                distance[i] = -1;
            }
        }
        return distance;
        
    }
};
